bam_sort.c

```diff
@@ -1916,16 +2010,131 @@ static inline int bam1_cmp_by_minhash(const bam1_tag a, const bam1_tag b)
         return bam1_cmp_core(a,b);
 }
 
+static template_coordinate_key_t* template_coordinate_key(bam1_t *b, template_coordinate_key_t *key, sam_hdr_t *hdr) {
+    uint8_t *data;
+
+    // defaults
+    key->tid1 = key->tid2 = INT32_MAX;
+    key->pos1 = key->pos2 = HTS_POS_MAX;
+    key->neg1 = key->neg2 = false;
+    key->mid  = "";
+
+    // update values
+    key->library = bam_get_library(hdr, b);
+    if (key->library == NULL) key->library = "";
+    key->name = bam_get_qname(b);
+    if (!(b->core.flag & BAM_FUNMAP)) { // read is mapped, update coordinates
+        key->tid1 = b->core.tid;
+        key->neg1 = bam_is_rev(b);
+        key->pos1 = (key->neg1) ? unclipped_end(b) : unclipped_start(b);
+    }
+    if (b->core.flag & BAM_FPAIRED && !(b->core.flag & BAM_FMUNMAP)) { // mate is mapped, update coordinates
+        char *cigar;
+        if ((data = bam_aux_get(b, "MC"))) {
+            if (!(cigar = bam_aux2Z(data))) {
+                fprintf(stderr, "[bam_sort] error: MC tag wrong type. Please use the MC tag provided by samtools fixmate.\n");
+                return NULL;
+            }
+        } else {
+            fprintf(stderr, "[bam_sort] error: no MC tag. Please run samtools fixmate on file first.\n");
+            return NULL;
+        }
+        key->tid2 = b->core.mtid;
+        key->neg2 = bam_is_mrev(b);
+        key->pos2 = (key->neg2) ? unclipped_other_end(b->core.mpos, cigar) : unclipped_other_start(b->core.mpos, cigar);
+    }
+
+    if ((data = bam_aux_get(b, "MI"))) {
+        if (!(key->mid=bam_aux2Z(data))) {
+            fprintf(stderr, "[bam_sort] error: MI tag wrong type (not a string).\n");
+            return NULL;
+        }
+        // remove training /1 or /2, for duplex sequencing data
+        int i;
+        for (i = strlen(key->mid) - 1; 0 <= i; --i) {
+            if (key->mid[i] == '/') {
+                key->mid[i] = '\0';
+                break;
+            }
+        }
```

Comment:
Editing `key->mid` here is a bad idea as it edits the pointed-to aux data in a way that can't be undone. Depending on where the `MI` tag is, it either silently removes everything after the `/`, or it results in a `Corrupted aux data` error.

You need to either take a copy of the `MI` tag, or to avoid more memory-related faff write a custom string comparison function that ignores the `/1` or `/2` on the end.

Reply:
I'm actually surprised that there's any need for trimming a `/1` or `/2` from an `MI` value. (In fact this code trims `/lots.of.words` too — anything from the last `/`.) There's no hint in _SAMtags.pdf_ that that is something that might be necessary.

If in practice you find that `MI` values are often just read names, including the `/1` or `/2`, and that you want to ignore that suffix for the purposes of picking out that unique ID for “All reads with the same MI tag represent the group of reads”, then this is something that the specification needs to address (or the tools doing this need fixing!). Please consider raising this as an hts-specs issue if so.

Reply:
Will do. For duplex sequencing, it is useful to distinguish which strand on the original source molecule the reads originate, so a /1 or /2 is used. I’ll add a spec issue later today.